docs/architecture/ADR-003-borrow-assets.md

```diff
@@ -0,0 +1,121 @@
+- Eventually, borrower pays repayment amount (in full or in part) using borrowed asset denomination
+- Borrower may withdraw collateral uTokens at any time as long as minimum overcollateralization requirement remains satisfied
+- On asset repayment amount = 0 for a given borrow position, facility forgets loan.
+- On uToken collateral amount = 0 for a given collateral position, facility forgets collateral.
+
+Additionally, the following events occur periodically (at the end of every block):
+- Repayment amounts for open borrow positions increases by a borrowed-token-specific interest rate
+- Open borrow positions are checked to see if they fall below liquidation threshold
+
+The `umee/x/leverage` module keeper will be responsible for remembering each open borrow or collateral position.
+If the same user account opens multiple of the same type of position (e.g. borrow) in the same token denomination, the second position simply increases the amount of the first.
+
+Additionally it has been discussed that, rather than an account's specific collateral positions being tied to specific borrow positions, the sum of all borrow and collateral positions related to an account is used to check the account's overcollateralization levels and mark its positions for liquidation.
+
+Note that the exchange rate of Assets:uAssets (e.g. `uatom:u/uatom`) will be a shifting exchange rate that grows with interest - see ADR-001.
+
+In contrast, the exchange rate of original assets:borrowed assets (e.g. `uatom:ether`) can only be determined using a price oracle.
+
+The true borrow exchange rate used in the transaction (e.g. `u/uatom:ether`) is derived from the two above.
+
+Note also that as a consequence of uToken interest, the asset value of uToken collateral increases over time, meaning a user who repays positions in full and redeems collateral uTokens will receive back more base assets than they deposited originally, reducing effective interest.
+
+## Detailed Design
+
+For the purposes of borrowing and repaying assets, as well as depositing and withdrawing uToken collateral, the `umee/x/leverage` module does not mint or burn tokens. It uses its module account to store collateral uTokens, its keeper to remember open positions, and the `x/bank` module to perform all necessary balance checks and token transfers.
+
+### Basic Message Types
+
+To implement the borrow/repay/collateral functionality of the Asset Facility, the four common message types will be:
+```go
+// MsgDepositCollateral - a user wishes to deposit uToken collateral of one or more types
+type MsgDepositCollateral struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
+}
+// MsgWithdrawCollateral - a user wishes to withdraw uToken collateral of one or more types
+type MsgWithdrawCollateral struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
+}
```

Comment:
Why do we need explicit messages for depositing and withdrawing collateral?

In compound, when a user deposits a Token, they are (optionally) able to enable it as "collateral" via a separate message. Maybe we add a `EnableCollateral` field to `MsgLendAsset` to avoid requiring a separate message?

If enabled, borrowing (up to the borrow limit) is is a one-step message.

In fact, I highly recommend you fiddle with https://app.compound.finance/ on the Rinkeby testnet. Send me your ETH address and I'll send you some ETH tokens.

@brentxu does this make sense?


![image](https://user-images.githubusercontent.com/5265989/135129272-4fd248ef-ebc9-44d1-a284-2142bcfbcb23.png)

Reply:
(This is the big change that will be made in an upcoming commit)

Reply:
Change is now made - marking as resolved